File: android/app/src/main/java/com/navilive/android/data/routing/RouteQualityAnalyzer.kt
package com.navilive.android.data.routing

import com.navilive.android.model.GeoPoint
import com.navilive.android.model.RouteStep
import com.navilive.android.model.SharedProductRules
import com.navilive.android.ui.RouteProjectionCore
import java.text.Normalizer
import java.util.Locale

/** Konkretna przyczyna, dla której asystent zaleca ręczne sprawdzenie trasy. */
internal enum class RouteQualityIssue {
    UnnamedTurn,
    RepeatedInstruction,
    MissingManeuverData,
    CloseOppositeManeuvers,
    ManeuverGeometryMismatch,
    IncompleteGeometry,
}

/** Ogólny poziom wyniku lokalnej analizy. Nie jest oceną bezpieczeństwa trasy. */
internal enum class RouteQualityLevel {
    Good,
    Review,
    InsufficientData,
}

/** Zalecenie dla interfejsu po analizie danych routera. */
internal enum class RouteQualityRecommendation {
    FollowCurrentGuidance,
    ReviewBeforeStarting,
    WaitForCompleteData,
}

/** Wynik ostrożnej, lokalnej analizy danych trasy. */
internal data class RouteQualityReport(
    val stepCount: Int,
    val unnamedTurnCount: Int,
    val repeatedInstructionCount: Int,
    val missingManeuverDataCount: Int,
    val closeOppositeManeuverCount: Int,
    val maneuverGeometryMismatchCount: Int,
    val hasIncompleteGeometry: Boolean,
    val qualityScore: Int,
    val confidencePercent: Int,
    val level: RouteQualityLevel,
) {
    /** Stabilna kolejność powoduje przewidywalny komunikat dla czytnika ekranu. */
    val issues: List<RouteQualityIssue>
        get() = buildList {
            if (unnamedTurnCount > 0) add(RouteQualityIssue.UnnamedTurn)
            if (repeatedInstructionCount > 0) add(RouteQualityIssue.RepeatedInstruction)
            if (missingManeuverDataCount > 0) add(RouteQualityIssue.MissingManeuverData)
            if (closeOppositeManeuverCount > 0) add(RouteQualityIssue.CloseOppositeManeuvers)
            if (maneuverGeometryMismatchCount > 0) add(RouteQualityIssue.ManeuverGeometryMismatch)
            if (hasIncompleteGeometry) add(RouteQualityIssue.IncompleteGeometry)
        }

    val issueCount: Int
        get() = unnamedTurnCount + repeatedInstructionCount + missingManeuverDataCount +
            closeOppositeManeuverCount +
            maneuverGeometryMismatchCount +
            if (hasIncompleteGeometry) 1 else 0

    val hasPotentialIssues: Boolean
        get() = level != RouteQualityLevel.Good

    val recommendation: RouteQualityRecommendation
        get() = when (level) {
            RouteQualityLevel.Good -> RouteQualityRecommendation.FollowCurrentGuidance
            RouteQualityLevel.Review -> RouteQualityRecommendation.ReviewBeforeStarting
            RouteQualityLevel.InsufficientData -> RouteQualityRecommendation.WaitForCompleteData
        }
}

/**
 * Analizuje tylko dane dostarczone przez router. Nie zastępuje routera i nie tworzy nowych
 * manewrów, dzięki czemu wynik może służyć jako wyjaśnialna pomoc dla użytkownika.
 */
internal object RouteQualityAnalyzer {
    private const val CLOSE_OPPOSITE_MANEUVER_DISTANCE_METERS = 60.0
    private const val MANEUVER_GEOMETRY_MAX_LATERAL_DISTANCE_METERS = 45.0
    private const val MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS = 25.0

    fun analyze(steps: List<RouteStep>, pathPoints: List<GeoPoint>): RouteQualityReport {
        val unnamedTurns = steps.count { step ->
            RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
                !step.maneuverType.equals("approach", ignoreCase = true) &&
                step.roadName?.trim().isNullOrEmpty()
        }
        val repeatedInstructions = steps.zipWithNext().count { (left, right) ->
            val leftInstruction = normalize(left.instruction)
            val rightInstruction = normalize(right.instruction)
            leftInstruction.isNotEmpty() &&
                leftInstruction == rightInstruction &&
                (left.distanceMeters <= 20 || right.distanceMeters <= 20)
        }
        val missingManeuverData = steps.count { step ->
            RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
                !step.maneuverType.equals("approach", ignoreCase = true) &&
                (step.instruction.isBlank() || step.maneuverPoint == null)
        }
        // Router może wstawić przejście dla pieszych między dwoma skrętami.
        // Porównujemy więc kolejne rzeczywiste manewry, a nie tylko sąsiednie wiersze.
        val maneuverSteps = steps.filter { step ->
            RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
                step.maneuverPoint != null
        }
        var previousManeuverStep: RouteStep? = null
        var previousManeuverDistance = 0.0
        var hasPreviousManeuverProjection = false
        var closeOppositeManeuvers = 0
        maneuverSteps.forEach { step ->
            val projection = step.maneuverPoint?.let { point ->
                val unconstrainedProjection = RouteProjectionCore.project(
                    pathPoints = pathPoints,
                    point = point,
                ) ?: return@let null
                if (unconstrainedProjection.lateralDistanceMeters >
                    MANEUVER_GEOMETRY_MAX_LATERAL_DISTANCE_METERS ||
                    hasPreviousManeuverProjection &&
                    unconstrainedProjection.distanceAlongRouteMeters +
                    MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS < previousManeuverDistance
                ) {
                    return@let null
                }
                RouteProjectionCore.project(
                    pathPoints = pathPoints,
                    point = point,
                    minimumDistanceAlongRouteMeters =
                        (previousManeuverDistance - MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS)
                            .coerceAtLeast(0.0),
                )
            }?.takeIf {
                it.lateralDistanceMeters <= MANEUVER_GEOMETRY_MAX_LATERAL_DISTANCE_METERS
            }
            if (projection == null) return@forEach
            val maneuverGap = projection.distanceAlongRouteMeters - previousManeuverDistance
            if (
                hasPreviousManeuverProjection &&
                previousManeuverStep != null &&
                areOppositeTurnModifiers(
                    previousManeuverStep.maneuverModifier,
                    step.maneuverModifier,
                ) &&
                maneuverGap >= -MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS &&
                maneuverGap <= CLOSE_OPPOSITE_MANEUVER_DISTANCE_METERS
            ) {
                closeOppositeManeuvers += 1
            }
            previousManeuverStep = step
            previousManeuverDistance = maxOf(
                previousManeuverDistance,
                projection.distanceAlongRouteMeters,
            )
            hasPreviousManeuverProjection = true
        }
        val hasIncompleteGeometry = steps.isEmpty() || pathPoints.size < 2 || pathPoints.any { point ->
            !point.latitude.isFinite() || !point.longitude.isFinite() ||
                point.latitude !in -90.0..90.0 || point.longitude !in -180.0..180.0
        }
        val maneuverGeometryMismatches = if (hasIncompleteGeometry) {
            0
        } else {
            countManeuverGeometryMismatches(steps, pathPoints)
        }
        val qualityScore = if (hasIncompleteGeometry) {
            0
        } else {
            (100 -
                unnamedTurns.coerceAtMost(3) * 12 -
                repeatedInstructions.coerceAtMost(2) * 12 -
                missingManeuverData.coerceAtMost(2) * 20 -
                closeOppositeManeuvers.coerceAtMost(2) * 15 -
                maneuverGeometryMismatches.coerceAtMost(2) * 18).coerceIn(0, 100)
        }
        val confidencePercent = if (hasIncompleteGeometry) {
            0
        } else {
            (100 -
                missingManeuverData.coerceAtMost(5) * 15 -
                maneuverGeometryMismatches.coerceAtMost(3) * 10).coerceIn(0, 100)
        }
        val level = when {
            hasIncompleteGeometry -> RouteQualityLevel.InsufficientData
            issuesFor(
                unnamedTurns = unnamedTurns,
                repeatedInstructions = repeatedInstructions,
                missingManeuverData = missingManeuverData,
                closeOppositeManeuvers = closeOppositeManeuvers,
                maneuverGeometryMismatches = maneuverGeometryMismatches,
                hasIncompleteGeometry = hasIncompleteGeometry,
            ).isNotEmpty() -> RouteQualityLevel.Review
            else -> RouteQualityLevel.Good
        }
        return RouteQualityReport(
            stepCount = steps.size,
            unnamedTurnCount = unnamedTurns,
            repeatedInstructionCount = repeatedInstructions,
            missingManeuverDataCount = missingManeuverData,
            closeOppositeManeuverCount = closeOppositeManeuvers,
            maneuverGeometryMismatchCount = maneuverGeometryMismatches,
            hasIncompleteGeometry = hasIncompleteGeometry,
            qualityScore = qualityScore,
            confidencePercent = confidencePercent,
            level = level,
        )
    }

    private fun issuesFor(
        unnamedTurns: Int,
        repeatedInstructions: Int,
        missingManeuverData: Int,
        closeOppositeManeuvers: Int,
        maneuverGeometryMismatches: Int,
        hasIncompleteGeometry: Boolean,
    ): List<RouteQualityIssue> = buildList {
        if (unnamedTurns > 0) add(RouteQualityIssue.UnnamedTurn)
        if (repeatedInstructions > 0) add(RouteQualityIssue.RepeatedInstruction)
        if (missingManeuverData > 0) add(RouteQualityIssue.MissingManeuverData)
        if (closeOppositeManeuvers > 0) add(RouteQualityIssue.CloseOppositeManeuvers)
        if (maneuverGeometryMismatches > 0) add(RouteQualityIssue.ManeuverGeometryMismatch)
        if (hasIncompleteGeometry) add(RouteQualityIssue.IncompleteGeometry)
    }

    /** Weryfikuje, czy punkty skrętów pasują do przebiegu trasy i występują po kolei. */
    private fun countManeuverGeometryMismatches(
        steps: List<RouteStep>,
        pathPoints: List<GeoPoint>,
    ): Int {
        val maneuverSteps = steps.filter { step ->
            RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
                !step.maneuverType.equals("approach", ignoreCase = true) &&
                step.maneuverPoint != null
        }
        if (maneuverSteps.isEmpty()) return 0

        var mismatches = 0
        var previousDistanceAlongRoute = 0.0
        var hasPreviousProjection = false
        maneuverSteps.forEach { step ->
            val point = step.maneuverPoint ?: run {
                mismatches += 1
                return@forEach
            }
            val unconstrainedProjection = RouteProjectionCore.project(
                pathPoints = pathPoints,
                point = point,
            )
            if (unconstrainedProjection == null ||
                unconstrainedProjection.lateralDistanceMeters >
                MANEUVER_GEOMETRY_MAX_LATERAL_DISTANCE_METERS
            ) {
                mismatches += 1
                return@forEach
            }
            if (hasPreviousProjection &&
                unconstrainedProjection.distanceAlongRouteMeters +
                MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS < previousDistanceAlongRoute
            ) {
                mismatches += 1
                return@forEach
            }
            val projection = RouteProjectionCore.project(
                pathPoints = pathPoints,
                point = point,
                minimumDistanceAlongRouteMeters =
                    (previousDistanceAlongRoute - MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS)
                        .coerceAtLeast(0.0),
            )
            if (projection == null || projection.lateralDistanceMeters > MANEUVER_GEOMETRY_MAX_LATERAL_DISTANCE_METERS) {
                mismatches += 1
                return@forEach
            }
            if (hasPreviousProjection &&
                projection.distanceAlongRouteMeters + MANEUVER_GEOMETRY_ORDER_TOLERANCE_METERS <
                previousDistanceAlongRoute
            ) {
                mismatches += 1
            }
            previousDistanceAlongRoute = maxOf(previousDistanceAlongRoute, projection.distanceAlongRouteMeters)
            hasPreviousProjection = true
        }
        return mismatches
    }

    private fun areOppositeTurnModifiers(left: String?, right: String?): Boolean {
        val normalizedLeft = left?.let(SharedProductRules.Instructions::normalizeModifier)
        val normalizedRight = right?.let(SharedProductRules.Instructions::normalizeModifier)
        val leftDirection = when {
            normalizedLeft?.contains("left") == true -> "left"
            normalizedLeft?.contains("right") == true -> "right"
            else -> null
        }
        val rightDirection = when {
            normalizedRight?.contains("left") == true -> "left"
            normalizedRight?.contains("right") == true -> "right"
            else -> null
        }
        return leftDirection != null && rightDirection != null && leftDirection != rightDirection
    }

    private fun normalize(value: String): String {
        return Normalizer.normalize(value, Normalizer.Form.NFD)
            .replace("\\p{M}+".toRegex(), "")
            .lowercase(Locale.ROOT)
            .replace("[^\\p{L}\\p{N}]+".toRegex(), " ")
            .replace("\\s+".toRegex(), " ")
            .trim()
    }
}

File: native-ios/NaviLive/Sources/Core/Navigation/RouteQualityAnalyzer.swift
import Foundation

/// Konkretna przyczyna, dla której asystent zaleca ręczne sprawdzenie trasy.
enum RouteQualityIssue: String, Equatable, Hashable {
  case unnamedTurn
  case repeatedInstruction
  case missingManeuverData
  case closeOppositeManeuvers
  case maneuverGeometryMismatch
  case incompleteGeometry
}

/// Ogólny poziom wyniku lokalnej analizy. Nie jest oceną bezpieczeństwa trasy.
enum RouteQualityLevel: Equatable, Hashable {
  case good
  case review
  case insufficientData
}

/// Zalecenie dla interfejsu po analizie danych routera.
enum RouteQualityRecommendation: Equatable, Hashable {
  case followCurrentGuidance
  case reviewBeforeStarting
  case waitForCompleteData
}

/// Wynik ostrożnej, lokalnej analizy danych trasy.
struct RouteQualityReport: Equatable {
  let stepCount: Int
  let unnamedTurnCount: Int
  let repeatedInstructionCount: Int
  let missingManeuverDataCount: Int
  let closeOppositeManeuverCount: Int
  let maneuverGeometryMismatchCount: Int
  let hasIncompleteGeometry: Bool
  let qualityScore: Int
  let confidencePercent: Int
  let level: RouteQualityLevel

  /// Stabilna kolejność powoduje przewidywalny komunikat dla VoiceOvera.
  var issues: [RouteQualityIssue] {
    var result: [RouteQualityIssue] = []
    if unnamedTurnCount > 0 { result.append(.unnamedTurn) }
    if repeatedInstructionCount > 0 { result.append(.repeatedInstruction) }
    if missingManeuverDataCount > 0 { result.append(.missingManeuverData) }
    if closeOppositeManeuverCount > 0 { result.append(.closeOppositeManeuvers) }
    if maneuverGeometryMismatchCount > 0 { result.append(.maneuverGeometryMismatch) }
    if hasIncompleteGeometry { result.append(.incompleteGeometry) }
    return result
  }

  var issueCount: Int {
    unnamedTurnCount + repeatedInstructionCount + missingManeuverDataCount + closeOppositeManeuverCount +
      maneuverGeometryMismatchCount +
      (hasIncompleteGeometry ? 1 : 0)
  }

  var hasPotentialIssues: Bool {
    level != .good
  }

  var recommendation: RouteQualityRecommendation {
    switch level {
    case .good:
      return .followCurrentGuidance
    case .review:
      return .reviewBeforeStarting
    case .insufficientData:
      return .waitForCompleteData
    }
  }
}

/// Analizuje tylko dane routera, bez tworzenia nowych manewrów i bez wysyłania danych do sieci.
enum RouteQualityAnalyzer {
  private static let closeOppositeManeuverDistanceMeters = 60.0
  private static let maneuverGeometryMaxLateralDistanceMeters = 45.0
  private static let maneuverGeometryOrderToleranceMeters = 25.0

  static func analyze(steps: [RouteStep], pathPoints: [GeoPoint]) -> RouteQualityReport {
    let unnamedTurns = steps.filter { step in
      RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
        step.maneuverType?.lowercased() != "approach" &&
        (step.roadName?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ?? true)
    }.count

    let repeatedInstructions = zip(steps, steps.dropFirst()).filter { left, right in
      let leftInstruction = normalize(left.instruction)
      let rightInstruction = normalize(right.instruction)
      return !leftInstruction.isEmpty &&
        leftInstruction == rightInstruction &&
        (left.distanceMeters <= 20 || right.distanceMeters <= 20)
    }.count
    let missingManeuverData = steps.filter { step in
      RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
        step.maneuverType?.lowercased() != "approach" &&
        (step.instruction.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ||
          step.maneuverPoint == nil)
    }.count
    // Router może wstawić przejście dla pieszych między dwoma skrętami.
    // Porównujemy więc kolejne rzeczywiste manewry, a nie tylko sąsiednie wiersze.
    let maneuverSteps = steps.filter { step in
      RouteStepSimplificationCore.isTurnLikeManeuver(step) && step.maneuverPoint != nil
    }
    var previousManeuverStep: RouteStep?
    var previousManeuverDistance = 0.0
    var hasPreviousManeuverProjection = false
    var closeOppositeManeuvers = 0
    for step in maneuverSteps {
      let projection: RouteProgressProjection? = {
        guard let point = step.maneuverPoint,
              let unconstrainedProjection = RouteProjectionCore.project(
                pathPoints: pathPoints,
                point: point
              ),
              unconstrainedProjection.lateralDistanceMeters <= maneuverGeometryMaxLateralDistanceMeters,
              !hasPreviousManeuverProjection ||
                unconstrainedProjection.distanceAlongRouteMeters + maneuverGeometryOrderToleranceMeters >= previousManeuverDistance,
              let projection = RouteProjectionCore.project(
                pathPoints: pathPoints,
                point: point,
                minimumDistanceAlongRouteMeters: max(
                  previousManeuverDistance - maneuverGeometryOrderToleranceMeters,
                  0
                )
              ),
              projection.lateralDistanceMeters <= maneuverGeometryMaxLateralDistanceMeters else {
          return nil
        }
        return projection
      }()
      guard let projection else { continue }
      let maneuverGap = projection.distanceAlongRouteMeters - previousManeuverDistance
      if hasPreviousManeuverProjection,
         let previousManeuverStep,
         areOppositeTurnModifiers(previousManeuverStep.maneuverModifier, step.maneuverModifier),
         maneuverGap >= -maneuverGeometryOrderToleranceMeters,
         maneuverGap <= closeOppositeManeuverDistanceMeters {
        closeOppositeManeuvers += 1
      }
      previousManeuverStep = step
      previousManeuverDistance = max(previousManeuverDistance, projection.distanceAlongRouteMeters)
      hasPreviousManeuverProjection = true
    }

    let hasIncompleteGeometry = steps.isEmpty || pathPoints.count < 2 || pathPoints.contains { point in
      !point.latitude.isFinite || !point.longitude.isFinite ||
        !(-90.0...90.0).contains(point.latitude) || !(-180.0...180.0).contains(point.longitude)
    }
    let maneuverGeometryMismatches = hasIncompleteGeometry
      ? 0
      : countManeuverGeometryMismatches(steps: steps, pathPoints: pathPoints)
    let qualityScore: Int
    if hasIncompleteGeometry {
      qualityScore = 0
    } else {
      qualityScore = max(
        0,
        min(
          100,
          100 -
            min(unnamedTurns, 3) * 12 -
            min(repeatedInstructions, 2) * 12 -
            min(missingManeuverData, 2) * 20 -
            min(closeOppositeManeuvers, 2) * 15 -
            min(maneuverGeometryMismatches, 2) * 18
        )
      )
    }
    let confidencePercent = hasIncompleteGeometry
      ? 0
      : max(
        0,
        min(
          100,
          100 -
            min(missingManeuverData, 5) * 15 -
            min(maneuverGeometryMismatches, 3) * 10
        )
      )
    let level: RouteQualityLevel
    if hasIncompleteGeometry {
      level = .insufficientData
    } else if unnamedTurns > 0 || repeatedInstructions > 0 || missingManeuverData > 0 ||
      closeOppositeManeuvers > 0 || maneuverGeometryMismatches > 0 {
      level = .review
    } else {
      level = .good
    }

    return RouteQualityReport(
      stepCount: steps.count,
      unnamedTurnCount: unnamedTurns,
      repeatedInstructionCount: repeatedInstructions,
      missingManeuverDataCount: missingManeuverData,
      closeOppositeManeuverCount: closeOppositeManeuvers,
      maneuverGeometryMismatchCount: maneuverGeometryMismatches,
      hasIncompleteGeometry: hasIncompleteGeometry,
      qualityScore: qualityScore,
      confidencePercent: confidencePercent,
      level: level
    )
  }

  /// Weryfikuje, czy punkty skrętów pasują do przebiegu trasy i występują po kolei.
  private static func countManeuverGeometryMismatches(
    steps: [RouteStep],
    pathPoints: [GeoPoint]
  ) -> Int {
    let maneuverSteps = steps.filter { step in
      RouteStepSimplificationCore.isTurnLikeManeuver(step) &&
        step.maneuverType?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() != "approach" &&
        step.maneuverPoint != nil
    }
    guard !maneuverSteps.isEmpty else { return 0 }

    var mismatches = 0
    var previousDistanceAlongRoute = 0.0
    var hasPreviousProjection = false
    for step in maneuverSteps {
      guard let point = step.maneuverPoint,
            let unconstrainedProjection = RouteProjectionCore.project(
              pathPoints: pathPoints,
              point: point
            ),
            unconstrainedProjection.lateralDistanceMeters <= maneuverGeometryMaxLateralDistanceMeters else {
        mismatches += 1
        continue
      }
      if hasPreviousProjection &&
        unconstrainedProjection.distanceAlongRouteMeters + maneuverGeometryOrderToleranceMeters < previousDistanceAlongRoute {
        mismatches += 1
        continue
      }
      guard let projection = RouteProjectionCore.project(
        pathPoints: pathPoints,
        point: point,
        minimumDistanceAlongRouteMeters: max(
          previousDistanceAlongRoute - maneuverGeometryOrderToleranceMeters,
          0
        )
      ),
      projection.lateralDistanceMeters <= maneuverGeometryMaxLateralDistanceMeters else {
        mismatches += 1
        continue
      }
      if hasPreviousProjection &&
        projection.distanceAlongRouteMeters + maneuverGeometryOrderToleranceMeters < previousDistanceAlongRoute {
        mismatches += 1
      }
      previousDistanceAlongRoute = max(previousDistanceAlongRoute, projection.distanceAlongRouteMeters)
      hasPreviousProjection = true
    }
    return mismatches
  }

  private static func areOppositeTurnModifiers(_ left: String?, _ right: String?) -> Bool {
    let leftValue = left.map { SharedProductRules.Instructions.normalizeModifier($0) }
    let rightValue = right.map { SharedProductRules.Instructions.normalizeModifier($0) }
    let leftDirection = leftValue?.contains("left") == true
      ? "left"
      : (leftValue?.contains("right") == true ? "right" : nil)
    let rightDirection = rightValue?.contains("left") == true
      ? "left"
      : (rightValue?.contains("right") == true ? "right" : nil)
    return leftDirection != nil && rightDirection != nil && leftDirection != rightDirection
  }

  private static func normalize(_ value: String) -> String {
    value
      .folding(options: [.diacriticInsensitive, .caseInsensitive], locale: Locale(identifier: "en_US_POSIX"))
      .replacingOccurrences(of: #"[^\p{L}\p{N}]+"#, with: " ", options: .regularExpression)
      .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
      .trimmingCharacters(in: .whitespacesAndNewlines)
  }
}
